cloudconfig/podcfg/podcfg.go

```diff
@@ -0,0 +1,464 @@
+// for a new juju caas pod. This is only relevant for the bootstrap pod.
+type BootstrapConfig struct {
+	instancecfg.BootstrapConfig
+}
+
+// ControllerConfig represents controller-specific initialization information
+// for a new juju caas pod. This is only relevant for controller pod.
+type ControllerConfig struct {
+	instancecfg.ControllerConfig
+}
+
+// AgentConfig returns an agent config.
+func (cfg *ControllerPodConfig) AgentConfig(
+	tag names.Tag,
+	toolsVersion version.Number,
+) (agent.ConfigSetterWriter, error) {
+	var password, cacert string
+	if cfg.Controller == nil {
+		password = cfg.APIInfo.Password
+		cacert = cfg.APIInfo.CACert
+	} else {
+		password = cfg.Controller.MongoInfo.Password
+		cacert = cfg.Controller.MongoInfo.CACert
+	}
+	configParams := agent.AgentConfigParams{
+		Paths: agent.Paths{
+			DataDir:         cfg.DataDir,
+			LogDir:          cfg.LogDir,
+			MetricsSpoolDir: cfg.MetricsSpoolDir,
+		},
+		Jobs:              cfg.Jobs,
+		Tag:               tag,
+		UpgradedToVersion: toolsVersion,
+		Password:          password,
+		Nonce:             cfg.MachineNonce,
+		APIAddresses:      cfg.APIHostAddrs(),
+		CACert:            cacert,
+		Values:            cfg.AgentEnvironment,
+		Controller:        cfg.ControllerTag,
+		Model:             cfg.APIInfo.ModelTag,
+	}
+	return agent.NewStateMachineConfig(configParams, cfg.Bootstrap.StateServingInfo)
+}
+
+// JujuTools returns the directory where Juju tools are stored.
+func (cfg *ControllerPodConfig) JujuTools() string {
+	return agenttools.SharedToolsDir(cfg.DataDir, cfg.AgentVersion())
+}
+
+// stateHostAddrs returns a list of mongo server addresses.
+func (cfg *ControllerPodConfig) stateHostAddrs() []string {
+	var hosts []string
+	if cfg.Bootstrap != nil {
+		hosts = append(hosts, net.JoinHostPort(
+			"localhost", strconv.Itoa(cfg.Bootstrap.StateServingInfo.StatePort)),
+		)
+	}
+	if cfg.Controller != nil {
+		hosts = append(hosts, cfg.Controller.MongoInfo.Addrs...)
+	}
+	return hosts
+}
+
+// APIHostAddrs returns a list of api server addresses.
+func (cfg *ControllerPodConfig) APIHostAddrs() []string {
+	var hosts []string
+	if cfg.Bootstrap != nil {
+		hosts = append(hosts, net.JoinHostPort(
+			"localhost", strconv.Itoa(cfg.Bootstrap.StateServingInfo.APIPort)),
+		)
+	}
+	if cfg.APIInfo != nil {
+		hosts = append(hosts, cfg.APIInfo.Addrs...)
+	}
+	return hosts
+}
+
+func (cfg *ControllerPodConfig) APIHosts() []string {
+	var hosts []string
+	if cfg.Bootstrap != nil {
+		hosts = append(hosts, "localhost")
+	}
+	if cfg.APIInfo != nil {
+		for _, addr := range cfg.APIInfo.Addrs {
+			host, _, err := net.SplitHostPort(addr)
+			if err != nil {
+				logger.Errorf("Can't split API address %q to host:port - %q", host, err)
+				continue
+			}
+			hosts = append(hosts, host)
+		}
+	}
+	return hosts
+}
+
+// AgentVersion returns the version of the Juju agent that will be configured
+// on the instance.
+func (cfg *ControllerPodConfig) AgentVersion() version.Binary {
```

Comment:
If this is even needed - don't do it in terms of tools.
I think we can delete this though as we can get by without it